src/EventListener/SessionListener.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2018 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\CoreBundle\Framework\ContaoFrameworkInterface;
+use Contao\CoreBundle\Routing\ScopeMatcher;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\EventListener\AbstractSessionListener;
+use Symfony\Component\HttpKernel\EventListener\SessionListener as BaseSessionListener;
+
+/**
+ * Decorates the default session listener.
+ *
+ * Symfony has recently changed their session listener to always make the
+ * response private if the session has been started. Although we agree with the
+ * change, it renders the HTTP cache unusable, because Contao always starts a
+ * session (e.g. to store the user's language). This listener circumvents
+ * Symfony's changes by not making the response private if the request is a
+ * Contao front end request.
+ *
+ * @author Leo Feyer <https://github.com/leofeyer>
+ */
+class SessionListener implements EventSubscriberInterface
+{
+    /**
+     * @var BaseSessionListener
+     */
+    private $inner;
+
+    /**
+     * @var ContaoFrameworkInterface
+     */
+    private $framework;
+
+    /**
+     * @var ScopeMatcher
+     */
+    private $scopeMatcher;
+
+    /**
+     * Constructor.
+     *
+     * @param BaseSessionListener      $inner
+     * @param ContaoFrameworkInterface $framework
+     * @param ScopeMatcher             $scopeMatcher
+     */
+    public function __construct(BaseSessionListener $inner, ContaoFrameworkInterface $framework, ScopeMatcher $scopeMatcher)
+    {
+        $this->inner = $inner;
+        $this->framework = $framework;
+        $this->scopeMatcher = $scopeMatcher;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function onKernelRequest(GetResponseEvent $event)
+    {
+        return $this->inner->onKernelRequest($event);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if (!$this->framework->isInitialized() || !$this->scopeMatcher->isFrontendMasterRequest($event)) {
+            $this->inner->onKernelResponse($event);
+
+            return;
+        }
+
+        $session = $event->getRequest()->getSession();
```

Comment:
@leofeyer There's an issue here. Either we have to call ` $this->inner->onKernelResponse($event);` at the end of the method again or we can leave out the whole `save()` part here because we're overriding the method.

Reply:
No, we must not call `$this->inner->onKernelResponse($event)` otherwise we have the same problem (private response) again.

Reply:
This does not make sense to me. If you do not forward to the decorated service, why do you call `save()` then?

Reply:
So the follow up listener, which checks `$session->isStarted()`, is not run.

Reply:
Which follow up listener? I think you confuse stuff here. The one that checks for this is the one you decorate, no?

Reply:
I'm not confusing anything. We have analyzed the process and set up the existing solution during the developer's meeting. Which issues does it cause?